overlord/servicestate/quota_control.go

```diff
@@ -296,3 +312,68 @@ func EnsureSnapAbsentFromQuota(st *state.State, snap string) error {
 	// the snap wasn't in any group, nothing to do
 	return nil
 }
+
+// QuotaChangeConflictError represents an error because of quota group conflicts between changes.
+type QuotaChangeConflictError struct {
+	Quota      string
+	ChangeKind string
+	// a Message is optional, otherwise one is composed from the other information
+	Message string
+}
+
+func (e *QuotaChangeConflictError) Error() string {
+	if e.Message != "" {
+		return e.Message
+	}
+	if e.ChangeKind != "" {
+		return fmt.Sprintf("quota group %q has %q change in progress", e.Quota, e.ChangeKind)
+	}
+	return fmt.Sprintf("quota group %q has changes in progress", e.Quota)
+}
+
+// CheckQuotaChangeConflictMany ensures that for the given quota groups no other
+// changes that alters them (like create, update, remove) are in
+// progress. If a conflict is detected an error is returned.
+func CheckQuotaChangeConflictMany(st *state.State, quotaNames []string) error {
+	quotaMap := make(map[string]bool, len(quotaNames))
+	for _, k := range quotaNames {
+		quotaMap[k] = true
+	}
+
+	for _, task := range st.Tasks() {
+		chg := task.Change()
+		if chg == nil || chg.IsReady() {
+			continue
+		}
+
+		quotas, err := affectedQuotas(task)
+		if err != nil {
+			return err
+		}
+
+		for _, quota := range quotas {
+			if quotaMap[quota] {
+				return &QuotaChangeConflictError{Quota: quota, ChangeKind: chg.Kind()}
+			}
+		}
+	}
+
+	return nil
+}
+
+func affectedQuotas(task *state.Task) ([]string, error) {
+	// so far only quota-control is relevant
+	if task.Kind() != "quota-control" {
+		return nil, nil
+	}
+
+	qcs := []QuotaControlAction{}
+	if err := task.Get("quota-control-actions", &qcs); err != nil {
+		return nil, fmt.Errorf("internal error: cannot get quota-control-action: %v", err)
```

Comment:
```suggestion
		return nil, fmt.Errorf("internal error: cannot get quota-control-actions: %v", err)
```

Reply:
I copy-pasted this, I suppose it's follow up material